### PR TITLE
DB backend rewrite, part 1

### PR DESCRIPTION
**Pull request by spbnick**
This starts the rewrite of the database code to transition to lightweight object-oriented model with centralized schema validation. There are plenty of things that can be done with it, and the main being clearer, more maintainable code. Other things include:

* convert database to YAML (code is ready now!);
* refactor the database structure to have data objects reflect the reality better (tree/suite/testcase/patterns/etc. objects);
* store trees in the YAML files explicitly, allowing to list them directly (instead of listing templates) and attach properties to them;
* expose database objects to Jinja templates directly (I *assume* that's possible), so that the templates could iterate over them and access their properties, instead of mushing the data together into just a few variables, and so we don't have to create functions for every possible data item we might need (e.g. waived state or maintainer addresses);
* actually support schema versioning and conversion, so that we could upgrade it smoothly without losing compatibility
* support database merging, so we could have two databases: public with the majority of tests, and private with those we cannot or are yet to publish

### CODE REVIEW

**Comment by wizardofoz123** [comment]
This looks like a lot of changes. Do you have the modified kpet-db somewhere? This looks like the yaml could be much less comprehensive to QE. I'm worried about building this many types (Struct, Node, List, ...), etc.

**Comment by spbnick** [comment]
Thanks for taking a look, Jakub!

Well, this is a redesign of a core part, with refactoring as a goal, and with a bit of the refactoring started. So it would be a considerable amount of changes, and I would like to do much more eventually. Still, I took the gradual change approach here, and I can squash a good deal of commits here, if that would be easier to review.

I don't have a modified kpet-db yet, and it's not needed so far - the code would work with the existing database, and we can still do a bunch of refactoring without changing it. I have a bunch of ideas for refactoring the database structure, though, and they're coming once we get past this. I need these changes here to be able to experiment easily, and to smoothly transition to a new schema, without a break in functionality.

As of now, there are no changes to the structure and the JSON/YAML, and so there would be no problem for QE, I want to make them as clear and as convenient as possible though. Once we get this in, as a side effect, we would be able to convert the JSON to YAML in the database.

Yes, the schema.py looks complicated. However, it is based on just a few principles:

* each class there is a complete schema (i.e. there is no separate schema class and type class)
* each schema can validate "data"
* "data" can be whatever you want, it's up to us how many or how little schemas we want, e.g. if we only ever want to have strings in our YAML files, then we don't need Float, Int, and Bool.
* each schema can "resolve" data - that is convert data to another representation (e.g. a filename to the contents of the corresponding YAML file, or a regex string to a compiled regex)
* each schema can be "recognized" (somewhat arbitrary term, I know) - that is, give us a schema for the "resolved" data, so that we can validate the resolved data, and continue resolving it and recognizing schema further (convert a filename to a YAML file, and then take a field from that file, and interpret it as a filename again, and so on).

This might be not that clear still, but we can work on improving it later. Right now it works well enough.

The fact that there are so many classes is just of the way I chose to organize it, and because of the number of type YAML files can represent. We don't have to take them further if we don't want to, and we can cut down on the number of types. We can say that we don't care if it's an Int or a String, we can just say Node(object) (or define a type doing that, e.g. "Object") and that would match anything.

What this gives us is the ability to, on one hand, link the YAML files together into a single hierarchy, no matter how many of them there are, and how the data is split, and on the other hand, assign classes to arbitrary nodes in the hierarchy, allowing us to encapsulate the data and attach methods for manipulation. This would let us change what we wish to put in separate files, and what we wish to keep in the same file. E.g. if we wish to have a testsuite in a single file, or if we want to have testcases in a separate file (or files). Or if we wish to have trees defined each in its own file, have a separate file for the trees, or have trees directly in the root file. That would only require a change to the schema in the corresponding class in data.py file. The code using those classes wouldn't have to care, and we wouldn't have to move all those `yaml.load()` invocations around.

**Comment by spbnick** [comment]
One of the first things I want to do, is to add a "tree" object to the YAML files, referring to their templates, so that we have trees listed explicitly and as objects, instead of e.g. listing the arbitrary template files. This would also let us assign further attributes to trees.

Merging separate kpet-db's (i.e. public and private) can then be accomplished by merging the list of testsuites and the list of trees. E.g. define a kpet.data.Base.merge() method accepting another kpet.data.Base object which would do that.

**Review by agustinhenze — COMMENTED** [review]
Hi nick, thanks for taking this huge jump on kpet, it really needs love but it's too much code for just one feature for me :/.

Leaving the subjective part aside ("too much code" is relative for everyone) I read what you wrote and makes sense. I understand the benefits of having a schema validation but I'd prefer using something more popular and already written like https://json-schema.org/ (openshift cli is using it) it works for both formats `json` and `yaml` or something similar.

I'm not going to block this PR but strongly recommend not merging this PR as is. I'd like to see what would be the equivalent implemented using json-schema for example and compare them. I think it wouldn't take more than 1 hour implementing it and we are going to inherit the documentation, other people is going to maintain that instead of us, etc hehe.

**Comment by spbnick** [comment]
Thank you for sharing your view, @agustinhenze. I appreciate the time you took to look at my code and read my explanations. I think your concerns about extra code and custom schema system are valid, and I was thinking about those too. I used both [json-schema](https://github.com/Scribery/tlog/blob/master/doc/schema.json) and [kwalify YAML schema](https://github.com/DIGImend/tablets/blob/master/schema.yml) before, as well as schemas for databases, XML, SGML, etc. I agree, using a well-documented and well-known schema and a well-tested validator would be great.

However, none of these would let us regroup files easily, wouldn't let us build a single hierarchy across multiple files and load it automatically, and wouldn't let us create classes at arbitrary points. While I consider having a schema a basic (necessary) requirement, it's these features that really get me excited, and I wouldn't let go of them easily. So much, that I don't mind having only basic schema validation (it's still better than none), and I intentionally created it light. And, while we can use an existing schema system/validator, *and* have these features on top, they won't integrate as easily, would result in more code, and we would have to write schemas separately and in another language too.

In short, I'd still like to play with this, I think it's gonna be fun and make things easier for us, but I'm open for more arguments, as I know I cannot foresee *everything* that this can potentially turn into.

**Comment by spbnick** [comment]
Another approach we could take is put all JSON/YAML into one file. Or, say, we agree that we won't need more than two layers of files and we know where the boundaries will be. Then we can limit ourselves just to one-two schemas and we would be able to cross file boundaries relatively easily. I still don't like the idea of writing all this by hand, but it would be more manageable. I don't think, though, that we're quite ready to make such decisions, that's why I thought we can have a lightweight "schema" and a flexible way of splitting data into files until we know our problem. And then, perhaps, we would be able to settle on an external schema validator and just a few fixed classes.

**Comment by wizardofoz123** [comment]
@spbnick Hi Nick, my review is similar to Agustin's. I'm not going to block this either.

If I understand correctly, then this is _not_ just about validation, but also data-shaping by using methods bound to them. It feels slightly over-engineered. Exactly what I've been told I sometimes do.
To me, a design like this is often more readable, but it is often _less_ readable for other people.

This looks like a framework to modify database structure by binding it to the object(s). Essentially ensuring, that whatever changes we need to do to the database, we can supplement by stacking objects and methods, this time using very little code, because this framework handles the "meat" of the work.
This basically turns kpet into a MacGyver for handling kpet-db data regardless of their organization.

**Comment by spbnick** [comment]
@wizardofoz123. Hi Jakub, thank you for your comment and review.

Yes, the validation is just a part of it (although it doesn't mean it's not important). Yes, it's more about having an easier time reshaping the data, as I'm sure we will need to. First for this, immediate refactoring, and then for the time when we'll have to handle something else than just pure Beaker XML. And I don't want us to be slowed down by having to rewrite those schemas and shuffle file loading around.

Yeah, I feel this is a little complex too (MacGyver indeed :smile:), and I certainly followed my creative urge here, which can lead to over-engineered and difficult code sometimes, but I think this complexity is well-contained, as it concerns data loading and validation only. We need the object-oriented representation of our data anyway, but if we'll want to get rid of the validation/loading (schema.py), we will just need to change the constructors (in data.py). I.e. this complexity barely affects the rest of the code, and the rest of the code doesn't need to know anything about it.

**Comment by spbnick** [comment]
I just found [JSON-LD](https://json-ld.org/), which seems to be addressing multi-file schema problem, and it has a [Python library](https://github.com/digitalbazaar/pyld). It's definitely more complex, it doesn't seem to support YAML (although there are ideas of making a similar one for YAML), but it might be something good enough, which is developed and maintained by somebody else. I'll take a closer look at it.

**Comment by spbnick** [comment]
Oh no, nope, I don't think we need JSON-LD, it's waaay too complicated. I'd rather use multiple per-file schemas.

**Comment by spbnick** [comment]
OK, I see nobody is much swayed by my argument :) So, as I said on the meeting, I'd like to offer a deal: I merge this, and if after two months you still think `schema.py` should be removed and we should do it in a different way, I'll do it. Meanwhile, I'll go on with refactoring the rest of kpet, and who knows, maybe I'll change my opinion myself :)

So, if nobody objects to this by tomorrow's evening, I'll merge this (after making CI pass), and we come back to it in two months.

**Review by wizardofoz123 — APPROVED** [review]
@spbnick: I actually don't disapprove. I'm just trying to take the same (mostly reasonable measures): avoid complexity and over-engineering where possible.
Feel free to merge this. Just don't forget to fix it or reach out to me if something breaks. :)
